File: Report/doc/LevenshteinRatio.hs
module Lib (levenshteinRatio) where

import Data.Array

levenshteinRatio :: String -> String -> Double
levenshteinRatio source target
  | source == target = 1.0
  | otherwise =
      let distance  = fromIntegral (calculateDistance source target)
          maxLength = fromIntegral (max (length source) (length target))
      in 1.0 - distance / maxLength

calculateDistance :: String -> String -> Int
calculateDistance source target = finalDistance sourceLength targetLength
  where
    sourceLength = length source
    targetLength = length target

    distanceMatrix = array
      ((0, 0), (sourceLength, targetLength))
      [
        ((i, j), distanceAtIndices i j) | i <- [0 .. sourceLength],
        j <- [0 .. targetLength]
      ]

    distanceAtIndices 0 j = j
    distanceAtIndices i 0 = i
    distanceAtIndices i j =
      minimum
        [ distanceMatrix ! (i - 1, j) + 1,
          distanceMatrix ! (i, j - 1) + 1,
          distanceMatrix ! (i - 1, j - 1) + cost i j
        ]

    cost i j
      | source !! (i - 1) == target !! (j - 1) = 0
      | otherwise = 1

    finalDistance i j = distanceMatrix ! (i, j)

File: Report/doc/LevenshteinRatio.cs
namespace LevenshteinRatio;

public static class Levenshtein {
  public static float Ratio(string target, string source) {
    if (source == target) {
      return 1;
    }

    if (string.IsNullOrEmpty(source) || string.IsNullOrEmpty(target)) {
      return 0;
    }

    var distance = new int[source.Length + 1, target.Length + 1];

    for (int i = 0; i <= source.Length; i++) {
      distance[i, 0] = i;
    }

    for (int i = 0; i <= target.Length; i++) {
      distance[0, i] = i;
    }

    for (int i = 1; i <= source.Length; i++) {
      for (int j = 1; j <= target.Length; j++) {
        int cost = target[j - 1] == source[i - 1] ? 0 : 1;

        distance[i, j] =
          Math.Min(
            Math.Min(
              distance[i - 1, j] + 1,
              distance[i, j - 1] + 1
            ),
            distance[i - 1, j - 1] + cost
          );
      }
    }

    var distanceAsFloat = Convert.ToSingle(distance[source.Length, target.Length]);

    return 1.0f - distanceAsFloat / Math.Max(source.Length, target.Length);
  }
}
